db/hints/sync_point.cc

```diff
@@ -135,12 +136,12 @@ sstring sync_point::encode() const {
     ser::serializer<sync_point_v1>::write(measure, v1);
 
     // Reserve version_size bytes for the version
-    bytes serialized{bytes::initialized_later{}, version_size + measure.size()};
-    seastar::simple_memory_output_stream out{reinterpret_cast<char*>(serialized.data()), serialized.size()};
+    sstring serialized{sstring::initialized_later{}, version_size + measure.size()};
+    seastar::simple_memory_output_stream out{serialized.data(), serialized.size()};
     ser::serializer<uint8_t>::write(out, 1);
     ser::serializer<sync_point_v1>::write(out, v1);
 
-    return base64_encode(serialized);
+    return base64_encode(bytes_view(reinterpret_cast<int8_t*>(serialized.data()), serialized.size()));
```

Comment:
`reinterpret_cast<int8_t*>(serialized.data()), serialized.size()));`
-- doesn't this break type aliasing rules (i.e. results in undefined behavior)?

From https://en.cppreference.com/w/cpp/language/reinterpret_cast
> Any object pointer type T1* can be converted to another object pointer type cv T2*. (...) In any case, the resulting pointer may only be dereferenced safely if allowed by the type aliasing rules (see below). 

And the type aliasing rules:
> Whenever an attempt is made to read or modify the stored value of an object of type DynamicType through a glvalue of type AliasedType, the behavior is undefined unless one of the following is true:
> 
> - AliasedType and DynamicType are similar.
> - AliasedType is the (possibly cv-qualified) signed or unsigned variant of DynamicType.
> - AliasedType is std::byte, (since C++17) char, or unsigned char: this permits examination of the object representation of any object as an array of bytes. 
> 
> Informally, two types are similar if, ignoring top-level cv-qualification:
> 
> - they are the same type; or
> - they are both pointers, and the pointed-to types are similar; or
> - they are both pointers to member of the same class, and the types of the pointed-to members are similar; or 

So if I understand correctly, `DynamicType` is `char` (`serialized.data()` is `char*`) and `AliasedType` is `int8_t`. And this case is not covered by the type aliasing rules.

Maybe I'm misreading something though.

Reply:
BTW. I have no idea why `bytes` is defined as `basic_sstring<int8_t, ...>` and not `basic_sstring<char, ...>` (or better, `basic_sstring<std::byte, ...>`. Then we wouldn't have this problem.

Reply:
Hm, good catch. So it looks like the conversion `bytes -> sstring_view` is safe, but `sstring -> bytes_view` is not. In that case, I suggest to:

- Stick to `bytes` as the base storage,
- Create a `sstring_view` object based on the bytes (via reinterpret casting),
- Use either the bytes or the string view where appropriate.

Reply:
If I understand the discussion in #14689 correctly, the current conversion is safe, but I'll wait for confirmation.

Reply:
If I understand it correctly, the conversion would be safe if `sstring` was using `unsigned char` or `byte` in its definition, it doesn't work for `char`. But I'm not sure, this is a high level of sorcery.
cc @michoecho please advise

Reply:
In any case, I would feel much safer if we didn't have to depend on this rule.

And Il think that the refactor to use sstring instead of bytes is not worth it, it introduces a lot of "noise" to the PR just to save a single `reinterpret_cast`.

Reply:
I confirm that this cast is technically illegal in the current revision of the C++ standard.
But also AFAIK it should not cause problems in any sane compiler.